scalanet/src/io/iohk/scalanet/peergroup/CloseableQueue.scala

```diff
@@ -0,0 +1,95 @@
+package io.iohk.scalanet.peergroup
+
+import cats.implicits._
+import cats.effect.concurrent.{TryableDeferred, Deferred}
+import monix.catnap.ConcurrentQueue
+import monix.eval.Task
+import monix.execution.{BufferCapacity, ChannelType}
+import scala.util.{Left, Right}
+
+/** Wraps an underlying concurrent queue so that polling can return None when
+  * the producer side is finished, or vice versa the producer can tell when
+  * the consumer is no longer interested in receiving more values.
+  *
+  *
+  * @param closed indicates whether the producer side has finished and whether
+  * the messages already in the queue should or discarded (true) or consumed (false).
+  * @param queue is the underlying message queue
+  */
+class CloseableQueue[A](
+    closed: TryableDeferred[Task, Boolean],
+    queue: ConcurrentQueue[Task, A]
+) {
+  import CloseableQueue.Closed
+
+  /** Fetch the next item from the queue, or None if the production has finished
+    * and the queue has been emptied.
+    */
+  def next(): Task[Option[A]] =
+    closed.tryGet.flatMap {
+      case Some(true) =>
+        Task.pure(None)
+
+      case Some(false) =>
+        queue.tryPoll
+
+      case None =>
+        Task.race(closed.get, queue.poll).flatMap {
+          case Left(_) =>
+            next()
+          case Right(item) =>
+            Task.pure(Some(item))
+        }
+    }
+
+  /** Stop accepting items in the queue. Clear items if `discard` is true, otherwise let them be drained.
+    * If the queue is already closed it does nothing; this is because either the producer or the consumer
+    * could have closed the queue before.
+    */
+  def close(discard: Boolean): Task[Unit] =
```

Comment:
i find booleans in public interfaces as something really confusing. Could't we provide two methods for this.?

Reply:
Sure, I'll add methods for it. I always use named arguments for this kind of stuff but I see your point.